### PR TITLE
WIP [cudaaligner] Use DefaultDeviceAllocator

### PR DESCRIPTION
**Pull request by ahehn-nv**

### CODE REVIEW

File: cudaaligner/src/myers_gpu.cu [hunk 1]
```diff
@@ -335,28 +336,33 @@ int32_t myers_compute_edit_distance(std::string const& target, std::string const
...
     cudaStream_t stream;
     CGA_CU_CHECK_ERR(cudaStreamCreate(&stream));
+    {
+        DefaultDeviceAllocator allocator;
 
-    int32_t max_sequence_length = std::max(get_size(target), get_size(query));
-    device_buffer<char> sequences_d(2 * max_sequence_length);
-    device_buffer<int32_t> sequence_lengths_d(2);
+        int32_t max_sequence_length = std::max(get_size(target), get_size(query));
+        device_buffer<char> sequences_d(2 * max_sequence_length);
```
Comment (tijyojwad) [inline cudaaligner/src/myers_gpu.cu]: any reason these `device_buffer` calls aren't using the allocator?

File: cudaaligner/src/myers_gpu.cu [hunk 2]
```diff
@@ -379,36 +385,42 @@ matrix<int32_t> myers_get_full_score_matrix(std::string const& target, std::stri
...
+
     cudaStream_t stream;
     CGA_CU_CHECK_ERR(cudaStreamCreate(&stream));
 
-    int32_t max_sequence_length = std::max(get_size(target), get_size(query));
-    device_buffer<char> sequences_d(2 * max_sequence_length);
-    device_buffer<int32_t> sequence_lengths_d(2);
+    {
+        int32_t max_sequence_length = std::max(get_size(target), get_size(query));
+        device_buffer<char> sequences_d(2 * max_sequence_length);
```
Comment (tijyojwad) [inline cudaaligner/src/myers_gpu.cu]: same, no allocator for these?

File: cudaaligner/include/claragenomics/cudaaligner/aligner.hpp [hunk 3]
```diff
@@ -78,11 +79,25 @@ class Aligner
 /// \param max_target_length Maximum length of target string
 /// \param max_alignments Maximum number of alignments to be performed
 /// \param type Type of aligner to construct
+/// \param allocator Allocator to use for internal device memory allocations
```
Comment (tijyojwad) [inline cudaaligner/include/claragenomics/cudaaligner/aligner.hpp]: for the default API, what do you think of hiding the DefaultDeviceAllocator in the implementation itself and not exposing an API with it? e.g. instantiating the DefaultDeviceAllocatin in `allocator.cpp` and passing that into the calls there

File: cudaaligner/src/hirschberg_myers_gpu.cu [hunk 4]
```diff
@@ -344,7 +344,7 @@ myers_compute_scores(
             }
             if (threadIdx.x == 0)
                 warp_carry = 0;
-            if (warp_mask == 0xffff'ffffu)
+            if (warp_mask == 0xffff'ffffu && (threadIdx.x == 0 || threadIdx.x == 31))
```
Comment (tijyojwad) [inline cudaaligner/src/hirschberg_myers_gpu.cu]: are these bug fixes to the myers implementation? did these cause deadlocks?

File: cudaaligner/include/claragenomics/cudaaligner/aligner.hpp (hunk 3 above)
Comment (tijyojwad) [reply cudaaligner/include/claragenomics/cudaaligner/aligner.hpp]: Sort of like how the other API doesn't expose the allocator itself, and only exposes `max_caching_size`

File: common/utils/src/cudautils.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace cudautils
+{
+
+std::size_t find_largest_contiguous_device_memory_section()
+{
+    // find the largest block of contiguous memory
+    size_t free;
+    size_t total;
+    cudaMemGetInfo(&free, &total);
```
[the remark this answers is not in the document]
Comment (ahehn-nv) [reply common/utils/src/cudautils.cpp:24]: fixed in PR #347

File: cudaaligner/src/myers_gpu.cu (hunk 1 above)
Comment (ahehn-nv) [reply cudaaligner/src/myers_gpu.cu:348]: Nope, I forgot about those. Thanks! I'll fix it.

File: cudaaligner/src/myers_gpu.cu (hunk 2 above)
Comment (ahehn-nv) [reply cudaaligner/src/myers_gpu.cu:395]: will fix.

File: cudaaligner/include/claragenomics/cudaaligner/aligner.hpp (hunk 3 above)
Comment (ahehn-nv) [reply cudaaligner/include/claragenomics/cudaaligner/aligner.hpp:82]: I think the API needs to expose both. I'll add an API call which only exposes `max_caching_size`.

File: cudaaligner/src/hirschberg_myers_gpu.cu (hunk 4 above)
Comment (ahehn-nv) [reply cudaaligner/src/hirschberg_myers_gpu.cu:347]: No, this was an oversight. It is not related to fixing the deadlocks I found.

**Comment by ahehn-nv** [comment]
superseded by #347 and #354

**Pull request closed without merging**